src/Compilers/CSharp/Portable/Parser/DirectiveParser.cs

```diff
@@ -671,30 +672,35 @@ private DirectiveTriviaSyntax ParseShebangDirective(SyntaxToken hash, SyntaxToke
 
         private SyntaxToken ParseEndOfDirectiveWithOptionalPreprocessingMessage()
         {
-            StringBuilder builder = null;
+            PooledStringBuilder builder = null;
 
-            if (this.CurrentToken.Kind != SyntaxKind.EndOfDirectiveToken &&
-                this.CurrentToken.Kind != SyntaxKind.EndOfFileToken)
+            // Skip the rest of the line until we hit a newline or EOF.  This follows the PP_Message portion of the specification.
+            var textWindow = this.lexer.TextWindow;
+            while (true)
             {
-                builder = new StringBuilder(this.CurrentToken.FullWidth);
-
-                while (this.CurrentToken.Kind != SyntaxKind.EndOfDirectiveToken &&
-                       this.CurrentToken.Kind != SyntaxKind.EndOfFileToken)
+                var ch = this.lexer.TextWindow.PeekChar();
+                if (ch is '\r' or '\n' || SyntaxFacts.IsNewLine(ch))
                 {
-                    var token = this.EatToken();
```

Comment:
this was a very odd impl choice for this method.  By eating tokens, it meant that we started interpreting the content as independent lexical constructs (which themselves could have their own trivia/spans/etc.).